src/rt/sched/backpressure.h

```diff
@@ -0,0 +1,249 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#pragma once
+
+/**
+ * ## Introduction
+ *
+ * The primary goal of backpressure is to prevent runaway growth of a cown's
+ * message queue that might eventually result in a Verona application running
+ * out of memory. Cowns that receive messages at a higher rate than they are
+ * able to process are considered "overloaded" and may lead to this unbounded
+ * queue growth. Once an overloaded cown is identified, cowns sending messages
+ * to it will be temporarily descheduled ("muted") so that the overloaded cown
+ * can work through its queue. Similarly, cowns sending to muted receivers will
+ * also be muted so that cowns do not themselves become overloaded as a
+ * consequence of being muted. Muted cowns will be rescheduled after the
+ * receiver that triggered their muting (the "mutor") is no longer muted or
+ * overloaded.
+ *
+ * ## Detecting Overloaded Cowns
+ *
+ * An overloaded cown is one that experiences message queue growth as it is
+ * processing messages. Since it would be prohibitively expensive to calculate
+ * the length of a message queue directly, a count of messages removed from the
+ * queue over some period is used as an approximation for message queue size.
+ * Like scheduler fairness, the period is determined using a token message that
+ * will occasionally fall out of the queue and then be inserted to begin the
+ * next period. The "load" of a cown at any time is the accumulation of messages
+ * processed over multiple periods.
+ *
+ * ## Muting Cowns
+ *
+ * All messages from one set of sending cowns to a set of receiving cowns are
+ * scanned to determine if the senders should be muted. The senders will be
+ * muted once their action completes if all of the following are true:
+ *   1. The receiver set does not contain any of the senders.
+ *   2. No sender is overloaded.
+ *   3. Any of the receivers are either overloaded or muted.
+ *
+ * A message is given "priority" if either condition 1 or 2 are false. A
+ * priority message will result in any muted receivers being unmuted so that the
+ * senders may make progress sooner.
+ *
+ * If all of the above conditions are met, the first receiver that is either
+ * overloaded or muted is identified as the mutor for the senders. First in this
+ * case means the least cown, determined by the mutlimessage cown order.
+ *
+ * ## Tracking Muted Cowns
+ *
+ * Once a scheduler thread completes a message action that would result in
```

Comment:
Is it reasonable to assume that someone reading this will be familiar with scheduler threads and how they work? On my first read this didn't stand out to me, now, I think it might be problematic. 

Theo, Do you think it would be distracting as part of the intro to cover a bit more about the scheduler threads related to backpressure? I think it might help make the tracking (and maybe other aspects) more understable. 

I'm not sure about that. Maybe what is needed is to introduce some background of "must know about scheduler threads to understanding mute tracking" right here at the start of this section.

Reply:
I think that the scheduler threads, mutli-message acquisition, and actions should be described in more detail elsewhere and should be a prerequisite for explaining backpressure. So I've added to the intro a short list of expectations for the reader.

Reply:
Letting people know what they need for background is an excellent approach.